pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -168,6 +143,62 @@ public class ModelInterpreter implements Serializable {
+                    Utils.logToTaskListener("Stage '${thisStage.name}' skipped due to earlier failure(s)")
+                    Utils.markStageSkippedForFailure(thisStage.name)
+                } else if (skipUnstable(root.options)) {
+                    Utils.logToTaskListener("Stage '${thisStage.name}' skipped due to earlier stage(s) marking the build as unstable")
+                    Utils.markStageSkippedForUnstable(thisStage.name)
+                } else {
+                    // While we run the top-level environment block after the top-level
+                    // agent, we do the reverse per-stage. Why? So that the per-stage
+                    // environment is populated before we evaluate any when condition,
+                    // and so that we don't go into a per-stage agent if the when condition
+                    // isn't satisfied.
+                    withEnvBlock(thisStage.getEnvVars(root, script, parentStage)) {
+                        if (evaluateWhen(thisStage.when)) {
+                            if (thisStage.parallelStages != null) {
+                                def parallelStages = [:]
+                                for (int i = 0; i < thisStage.parallelStages.stages.size(); i++) {
+                                    Stage parallelStage = thisStage.parallelStages.getStages().get(i)
+                                    parallelStages.put(parallelStage.name,
+                                    evaluateStage(root, thisStage.agent ?: parentAgent, parallelStage, firstError, thisStage))
+                                }
+                                script.parallel(parallelStages)
+                            } else {
+                                inDeclarativeAgent(thisStage, root, thisStage.agent) {
+                                    withCredentialsBlock(thisStage.environment) {
+                                        toolsBlock(thisStage.agent ?: root.agent, thisStage.tools) {
+                                            // Execute the actual stage and potential post-stage actions
+                                            executeSingleStage(root, thisStage, parentAgent)
+                                        }
+                                    }
+                                }
+                            }
+                        } else {
+                            Utils.logToTaskListener("Stage '${thisStage.name}' skipped due to when conditional")
+                            Utils.markStageSkippedForConditional(thisStage.name)
+                        }
+                    }
+                }
+            } catch (Exception e) {
+                script.getProperty("currentBuild").result = Result.FAILURE
+                Utils.markStageFailedAndContinued(thisStage.name)
+                if (firstError == null) {
+                    firstError = e
+                }
+            }
+
+            if (firstError != null) {
+                throw firstError
```

Comment:
So the nested stages were doing `post` fine, but the *parent* stage was not. So...juggling things around. Hooray.